### PR TITLE
Check if generator has next normal sample cache methods in normal_distribution

### LINKED ISSUE #39618: Fix `normal_distribution` check for RNG interface
[`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) should be replaced with SFINAE that checks if `RNG` typename has `next_double_normal_sample`, `set_next_double_normal_sample`, `next_float_normal_sample`, `set_next_float_normal_sample` methods

And [here](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L171) too

cc @pbelevich

### PR DESCRIPTION
**Pull request by pbelevich**
This change replaces [`#if !defined(__CUDACC__) && !defined(__HIPCC__)`](https://github.com/pytorch/pytorch/blob/856215509d89c935cd1768ce4b496d4fc0e919a6/aten/src/ATen/core/DistributionsHelper.h#L147) with SFINAE expression that checks if RNG typename has next_double_normal_sample, set_next_double_normal_sample, next_float_normal_sample, set_next_float_normal_sample methods

It is required by (and manually tested with) https://github.com/pytorch/csprng/pull/28

Fixes #39618

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#39816 Check if generator has next normal sample cache methods in normal_distribution**

Differential Revision: [D22002599](https://our.internmc.facebook.com/intern/diff/D22002599)

### CODE REVIEW

File: aten/src/ATen/core/DistributionsHelper.h
Comment (ezyang) [inline aten/src/ATen/core/DistributionsHelper.h:129]: This is a really big macro. Is there really no way to do this without macro?

**Comment by ezyang** [comment]
@smessmer would you mind helping review the C++ portions of this?

File: aten/src/ATen/core/DistributionsHelper.h
Comment (pbelevich) [reply aten/src/ATen/core/DistributionsHelper.h:129]: hm... this macro is 8 loc

Comment (ezyang) [reply aten/src/ATen/core/DistributionsHelper.h:129]: Apparently, this macro is a version of an old-style detection idiom in C++. In C++17 you could have just use `std::is_detected` to do this, no macros necessary. The way we generally prefer to backport features like this is to backport an implementation of the macro in question (e.g., the reference implementation from cppreference) so that we can easily switch to using the standard library version when we eventually upgrade to C++.

File: aten/src/ATen/core/DistributionsHelper.h
Comment (ezyang) [inline aten/src/ATen/core/DistributionsHelper.h:210]: This looks wrong. T is double I should never try to get the next float sample, right?

**Comment by ezyang** [comment]
This all seems a bit finely balanced for my taste. In particular, if I misspell the method name for getting/setting the cache values, I will just silently start wasting random numbers with no indication from the compiler at all.

You have control over all of the classes in question. Can't you do a more direct type test, e.g., forward declare some generator variants and just test if the type is one of those generator types before using the method names?

File: aten/src/ATen/core/DistributionsHelper.h
Comment (smessmer) [inline aten/src/ATen/core/DistributionsHelper.h:130]: This can be more easily implemented with `guts::void_t`, something like

```cpp
template<class T, class Enable = void> struct has_member_X : std::false_type {};
template<class T> struct has_member_X<T, guts::void_t<T::X>> : std::true_type {};
```

File: aten/src/ATen/core/DistributionsHelper.h
Comment (smessmer) [inline aten/src/ATen/core/DistributionsHelper.h:148]: nit: use `enable_if_t` and you don't need the `typename` or the `::type`

File: aten/src/ATen/core/DistributionsHelper.h
Comment (smessmer) [inline aten/src/ATen/core/DistributionsHelper.h:161]: Instead of SFINAE, you could use `guts::if_constexpr` here to combine both into one function and reduce the amount of code. same for the setter function below.

File: aten/src/ATen/core/DistributionsHelper.h
Comment (pbelevich) [reply aten/src/ATen/core/DistributionsHelper.h:210]: addressed

File: aten/src/ATen/core/DistributionsHelper.h
Comment (pbelevich) [reply aten/src/ATen/core/DistributionsHelper.h:129]: If you' re saying that I have to implement `std::is_detected` then can I do it in follow up PR and leave this as is?

File: aten/src/ATen/core/DistributionsHelper.h
Comment (pbelevich) [reply aten/src/ATen/core/DistributionsHelper.h:154]: not possible for now, because of "RuntimeError: CUDA error: unspecified launch failure"

File: aten/src/ATen/core/DistributionsHelper.h
Comment (pbelevich) [reply aten/src/ATen/core/DistributionsHelper.h]: addressed

**Comment by pbelevich** [comment]
> This all seems a bit finely balanced for my taste. In particular, if I misspell the method name for getting/setting the cache values, I will just silently start wasting random numbers with no indication from the compiler at all.
> 
> You have control over all of the classes in question. Can't you do a more direct type test, e.g., forward declare some generator variants and just test if the type is one of those generator types before using the method names?

@ezyang in future - yes, this PR is just to unblock CSPRNG release

**Review by smessmer — APPROVED** [review]
Looks good to me now. Please make sure @ezyang's concerns are addressed before landing.

**Comment by ezyang** [comment]
> in future - yes, this PR is just to unblock CSPRNG release

OK, in that case, write into the code what the intended refactor is so that we don't forget about it.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**
